# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: netwerk/ipc/ParentProcessDocumentChannel.cpp
# Commit: 7f7b98339065
# Full Hash: 7f7b983390650cbc7d736e92fd3e1f629a30ac02
# Author: Jean-Yves Avenard <jyavenard@mozilla.com>
# Date: 2020-06-09 15:16:49
# Regressor Bug: 1641737
# File Overlap Count: 1
# Description:
#   Bug 1641737 - P7. Make the DLL and PPDC promises use direct task dispatch. r=mattwoodrow
#   
#   We attempt to reduce the number of event loop iterations, which would bring us back to a similar behaviour to the pre-DocumentChannel days.
#   
#   This prevent some tests to have increase intermittent failures.
# ==============================================================================

diff -r 32fba417ebd0 -r 7f7b98339065 netwerk/ipc/ParentProcessDocumentChannel.cpp
--- a/netwerk/ipc/ParentProcessDocumentChannel.cpp	Tue Jun 09 07:24:15 2020 +0000
+++ b/netwerk/ipc/ParentProcessDocumentChannel.cpp	Tue Jun 09 07:24:23 2020 +0000
@@ -66,6 +66,9 @@
   mStreamFilterEndpoints = std::move(aStreamFilterEndpoints);
 
   RefPtr<RedirectToRealChannelPromise> p = mPromise.Ensure(__func__);
+  // We make the promise use direct task dispatch in order to reduce the number
+  // of event loops iterations.
+  mPromise.UseDirectTaskDispatch(__func__);
 
   nsresult rv =
       gHttpHandler->AsyncOnChannelRedirect(this, channel, aRedirectFlags);
@@ -176,7 +179,7 @@
                 ->Then(
                     GetCurrentThreadSerialEventTarget(), __func__,
                     [self](RedirectToRealChannelPromise::ResolveOrRejectValue&&
-                               aValue) {
+                               aValue) -> RefPtr<RedirectToRealChannelPromise> {
                       MOZ_ASSERT(aValue.IsResolve());
                       nsresult rv = aValue.ResolveValue();
                       if (NS_FAILED(rv)) {
@@ -186,8 +189,12 @@
                       self->mListener = nullptr;
                       self->mCallbacks = nullptr;
                       self->RemoveObserver();
-                      return RedirectToRealChannelPromise::
-                          CreateAndResolveOrReject(std::move(aValue), __func__);
+                      auto p =
+                          MakeRefPtr<RedirectToRealChannelPromise::Private>(
+                              __func__);
+                      p->UseDirectTaskDispatch(__func__);
+                      p->ResolveOrReject(std::move(aValue), __func__);
+                      return p;
                     });
         // We chain the promise the DLL is waiting on to the one returned by
         // RedirectToRealChannel. As soon as the promise returned is
